### DISCUSSION TITLE
What is the difference between treemap and opt?

### DISCUSSION THREAD

**Discussion by marcuswang6** [jax-ml/jax#10514] [Q&A]
**_For params 'p', use sample x and y to update it, what's the difference between following codes?_**
`grads = jax.grad(loss)(p, x, y)`
`updates, state = opt.update(grads, state)`
`p = optax.apply_updates(p, updates)`

`grads = jax.grad(loss)(p, x, y)`
`inner_sgd_fn = lambda g, state: (state - alpha*g)`
`p = jax.tree_util.tree_multimap(inner_sgd_fn, grads, p)`

**Comment by YouJiacheng** [comment]
optax internally use treemap, but provides a lot of gradient transform to implement advanced optimizer beyond SGD.
